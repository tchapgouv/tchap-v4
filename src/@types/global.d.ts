/*
Copyright 2020-2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import "matrix-react-sdk/src/@types/global"; // load matrix-react-sdk's type extensions first
import type { Renderer } from "react-dom";
import type { logger } from "matrix-js-sdk/src/logger";

type ElectronChannel =
    | "app_onAction"
    | "before-quit"
    | "check_updates"
    | "install_update"
    | "ipcCall"
    | "ipcReply"
    | "loudNotification"
    | "preferences"
    | "seshat"
    | "seshatReply"
    | "setBadgeCount"
    | "update-downloaded"
    | "userDownloadCompleted"
    | "userDownloadAction"
    | "openDesktopCapturerSourcePicker"
    | "userAccessToken"
    | "serverSupportedVersions";

declare global {
    interface Window {
        mxSendRageshake: (text: string, withLogs?: boolean) => void;
        matrixLogger: typeof logger;
        matrixChat: ReturnType<Renderer>;

        // electron-only
        electron?: Electron;

        // opera-only
        opera?: any;

        // https://developer.mozilla.org/en-US/docs/Web/API/InstallTrigger
        InstallTrigger: any;
    }

    interface Electron {
        on(channel: ElectronChannel, listener: (event: Event, ...args: any[]) => void): void;
        send(channel: ElectronChannel, ...args: any[]): void;
    }
}

// add method which is missing from the node typing
declare module "url" {
    interface Url {
        format(): string;
    }
}
