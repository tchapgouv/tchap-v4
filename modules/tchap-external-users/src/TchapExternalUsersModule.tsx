/*
Copyright 2022 New Vector Ltd. t/a Element

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { RuntimeModule } from "@matrix-org/react-sdk-module-api/lib/RuntimeModule";
import { ModuleApi } from "@matrix-org/react-sdk-module-api/lib/ModuleApi";
import { CantAddDialog  } from "./components/CantAddDialog";
import { AddRoomListener, ExternalUsersLifecycle } from "@matrix-org/react-sdk-module-api/lib/tchap/lifecycles/ExternalUsers";
import { DialogProps } from "@matrix-org/react-sdk-module-api/lib/components/DialogContent";

import React from "react";

export default class TchapExternalUsersModule extends RuntimeModule {
    private domain: String;

    public constructor(moduleApi: ModuleApi) {
        super(moduleApi);

        // const rawConfig = this.moduleApi.getConfigValue(
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_NAMESPACE,
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_KEY,
        //   );
      
        //   assertValidGuestModuleConfig(rawConfig);
        //   this.config = rawConfig;

        this.on(ExternalUsersLifecycle.AddRoom, this.onAddRoom);
    }

    protected onAddRoom: AddRoomListener = async () => {
        await this.moduleApi.openDialog<null, DialogProps, CantAddDialog>("test", (props, ref) => <CantAddDialog ref={ref} {...props} />);
    };
}
