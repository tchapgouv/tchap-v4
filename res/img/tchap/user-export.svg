<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M19.5002 11.5094C19.8949 11.5094 20.281 11.4736 20.6555 11.4051C20.7505 11.9558 20.8 12.5221 20.8 13.1C20.8 18.5676 16.3676 23 10.9 23C5.43238 23 1 18.5676 1 13.1C1 7.63238 5.43238 3.2 10.9 3.2C11.7973 3.2 12.6666 3.31936 13.4931 3.54308C13.3177 4.11118 13.2234 4.71396 13.2234 5.3384C13.2234 8.74655 16.0336 11.5094 19.5002 11.5094ZM11.1308 21.4653C13.2927 21.4653 15.2551 20.6258 16.6993 19.2603C15.7939 17.0769 13.6417 15.5412 11.1308 15.5412C8.61988 15.5412 6.46768 17.0769 5.56225 19.2603C7.00643 20.6258 8.96888 21.4653 11.1308 21.4653ZM11.2799 7.81768C9.78854 7.95149 8.62035 9.18434 8.62035 10.6854C8.62035 12.2759 9.93179 13.5652 11.5495 13.5652C12.6276 13.5652 13.5696 12.9927 14.0781 12.1401C12.6655 11.1035 11.6462 9.57751 11.2799 7.81768Z" fill="currentColor"/>
<path d="M18.5033 0.999999C18.0923 0.999999 17.7502 1.34413 17.7502 1.74465L17.7502 5.79526L17.8105 7.01059L17.3115 6.41134L16.1704 5.1862C16.0409 5.04077 15.86 4.96364 15.6702 4.96364C15.306 4.96364 15 5.2367 15 5.61876C15 5.81535 15.0788 5.96079 15.2149 6.09806L17.9357 8.73968C18.1269 8.93251 18.3026 9 18.5033 9C18.6975 9 18.8716 8.93251 19.0709 8.73968L21.7918 6.09806C21.9264 5.96079 22 5.81535 22 5.61876C22 5.2367 21.7074 4.96364 21.3365 4.96364C21.1534 4.96364 20.9591 5.04077 20.8296 5.1862L19.6953 6.41134L19.1895 7.01059L19.2498 5.79526L19.2498 1.74465C19.2498 1.34413 18.9144 0.999999 18.5033 0.999999Z" fill="currentColor"/>
</svg>
