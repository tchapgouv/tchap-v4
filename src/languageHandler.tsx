/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import {
    IVariables,
    TranslatedString,
    TranslationKey as ReactTranslationKey,
    // eslint-disable-next-line camelcase
    _t as react_t,
    // eslint-disable-next-line camelcase
    _td as react_td,
    // eslint-disable-next-line camelcase
    _tDom as react_tDom,
    Tags,
    UserFriendlyError as ReactUserFriendlyError,
    ErrorOptions,
} from "matrix-react-sdk/src/languageHandler";
import { Leaves } from "matrix-react-sdk/src/@types/common";

import type ReactEN from "matrix-react-sdk/src/i18n/strings/en_EN.json";
import type EN from "./i18n/strings/en_EN.json";

/**
 * This module wraps languageHandler in the matrix-react-sdk and adds type casts to include translations
 * which we know will be injected by webpack.
 */

export type TranslationKey = Leaves<typeof EN & typeof ReactEN, "|", string | { other: string }, 4>;

export class UserFriendlyError extends ReactUserFriendlyError {
    public constructor(message: TranslationKey, substitutionVariablesAndCause?: IVariables & ErrorOptions) {
        super(message as ReactTranslationKey, substitutionVariablesAndCause);
    }
}

export function _td(s: TranslationKey): TranslationKey {
    return react_td(s as ReactTranslationKey);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _t(text: TranslationKey, variables?: IVariables): string;
export function _t(text: TranslationKey, variables: IVariables | undefined, tags: Tags): React.ReactNode;
export function _t(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_t(text as ReactTranslationKey, variables, tags!);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _tDom(text: TranslationKey, variables?: IVariables): TranslatedString;
export function _tDom(text: TranslationKey, variables: IVariables, tags: Tags): React.ReactNode;
export function _tDom(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_tDom(text as ReactTranslationKey, variables!, tags!);
}
