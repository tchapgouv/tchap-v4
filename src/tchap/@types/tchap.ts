export enum TchapRoomType {
    Direct = "direct",
    Private = "private",
    External = "external",
    Forum = "forum",
    Unknown = "unknown",
}

export enum TchapRoomAccessRule {
    Unrestricted = "unrestricted", // accessible to externals
    Restricted = "restricted", // not accessible to externals
}

export interface TchapIAccessRuleEventContent {
    rule: TchapRoomAccessRule; // eslint-disable-line camelcase
}

export const TchapRoomAccessRulesEventId = "im.vector.room.access_rules";
