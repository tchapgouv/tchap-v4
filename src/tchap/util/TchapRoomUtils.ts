/**
 * Tchap Room utils.
 */

import { Room } from "matrix-js-sdk/src/matrix";
import { MatrixClientPeg } from "matrix-react-sdk/src/MatrixClientPeg";
import { ISendEventResponse } from "matrix-js-sdk/src/@types/requests";
import * as utils from "matrix-js-sdk/src/utils";
import { Method } from "matrix-js-sdk/src/http-api";

import { TchapRoomAccessRule, TchapRoomAccessRulesEventId, TchapRoomType } from "../@types/tchap";

export default class TchapRoomUtils {
    //inspired by https://github.com/tchapgouv/tchap-android/blob/develop/vector/src/main/java/fr/gouv/tchap/core/utils/RoomUtils.kt#L31
    //direct type is not handled yet
    static getTchapRoomType(room: Room): TchapRoomType {
        const isEncrypted: boolean = this.isRoomEncrypted(room.roomId);
        const tchapRoomAccessRule: TchapRoomAccessRule = this.getTchapRoomAccessRule(room);
        return this.getTchapRoomTypeInternal(isEncrypted, tchapRoomAccessRule);
    }

    static getTchapRoomTypeInternal(isEncrypted: boolean, tchapRoomAccessRule: TchapRoomAccessRule): TchapRoomType {
        if (!isEncrypted) {
            return TchapRoomType.Forum;
        }
        if (!tchapRoomAccessRule) {
            return TchapRoomType.Unknown;
        }
        if (tchapRoomAccessRule === TchapRoomAccessRule.Restricted) {
            return TchapRoomType.Private;
        }
        if (tchapRoomAccessRule === TchapRoomAccessRule.Unrestricted) {
            return TchapRoomType.External;
        }
        return TchapRoomType.Unknown;
    }

    /**
     * Get Tchap Room Access Rule
     * @param room
     * @returns string that matches of one TchapRoomAccessRule //todo or null? or empty?
     */
    static getTchapRoomAccessRule(room: Room): TchapRoomAccessRule {
        return room.currentState.getStateEvents(TchapRoomAccessRulesEventId, "")?.getContent().rule;
    }

    /**
     * Is room encrypted? search in local storage or in server if needed
     * @param roomId
     * @returns true if room is encrypted, false if not
     */
    static isRoomEncrypted(roomId: string): boolean {
        return MatrixClientPeg.get().isRoomEncrypted(roomId);
    }

    /**
     * Update public room join rule to "invite" state (so that the room be private).
     * @returns Promise which resolves: to an ISendEventResponse object
     * @returns Rejects: with an error response.
     */
    static updateRoomJoinRuleToInvite(
        roomId: string,
    ): Promise<ISendEventResponse> {
        const client = MatrixClientPeg.get();
        const content = { join_rule: 'invite' };
        const path = utils.encodeUri("/rooms/$roomId/state/m.room.join_rules", {
            $roomId: roomId,
        });

        return client.http.authedRequest(Method.Put, path, undefined, content);
    }

    /**
     * Turn on the encryption in a room (if this is not already done).
     * @returns Promise which resolves: to an ISendEventResponse object
     * @returns Rejects: with an error response.
     */
    static turnOnEncryptionInThisRoom(
        roomId: string,
    ): Promise<ISendEventResponse> {
        const client = MatrixClientPeg.get();
        const content = { algorithm: 'm.megolm.v1.aes-sha2' };
        const path = utils.encodeUri("/rooms/$roomId/state/m.room.encryption", {
            $roomId: roomId,
        });

        return client.http.authedRequest(Method.Put, path, undefined, content);
    }

    /**
     * Remove a room from the room directory.
     * @returns Promise which resolves: to an ISendEventResponse object
     * @returns Rejects: with an error response.
     */
    static removeRoomFromRoomDirectory(
        roomId: string,
    ): Promise<ISendEventResponse> {
        const client = MatrixClientPeg.get();
        const content = { visibility: 'private' };
        const path = utils.encodeUri("/rooms/$roomId/state/m.room.history_visibility", {
            $roomId: roomId,
        });

        return client.http.authedRequest(Method.Put, path, undefined, content);
    }
}
