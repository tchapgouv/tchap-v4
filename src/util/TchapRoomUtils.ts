/**
 * Tchap Room utils.
 */

import { Room } from "matrix-js-sdk/src/matrix";
import { MatrixClientPeg } from "matrix-react-sdk/src/MatrixClientPeg";

import { TchapRoomAccessRule, TchapRoomAccessRulesEventId, TchapRoomType } from "../tchap/@types/tchap";

export default class TchapRoomUtils {
    //inspired by https://github.com/tchapgouv/tchap-android/blob/develop/vector/src/main/java/fr/gouv/tchap/core/utils/RoomUtils.kt#L31
    //direct type is not handled yet
    static getTchapRoomType(room: Room): TchapRoomType {
        const isEncrypted: boolean = this.isRoomEncrypted(room.roomId);
        const tchapRoomAccessRule: TchapRoomAccessRule = this.getTchapRoomAccessRule(room);
        return this.getTchapRoomTypeInternal(isEncrypted, tchapRoomAccessRule);
    }

    static getTchapRoomTypeInternal(isEncrypted: boolean, tchapRoomAccessRule: TchapRoomAccessRule): TchapRoomType {
        if (!isEncrypted) {
            return TchapRoomType.Forum;
        }
        if (!tchapRoomAccessRule) {
            return TchapRoomType.Unknown;
        }
        if (tchapRoomAccessRule === TchapRoomAccessRule.Restricted) {
            return TchapRoomType.Private;
        }
        if (tchapRoomAccessRule === TchapRoomAccessRule.Unrestricted) {
            return TchapRoomType.External;
        }
        return TchapRoomType.Unknown;
    }

    /**
     * Get Tchap Room Access Rule
     * @param room
     * @returns string that matches of one TchapRoomAccessRule //todo or null? or empty?
     */
    static getTchapRoomAccessRule(room: Room): TchapRoomAccessRule {
        return room.currentState.getStateEvents(TchapRoomAccessRulesEventId, "")?.getContent().rule;
    }

    /**
     * Is room encrypted? search in local storage or in server if needed
     * @param roomId
     * @returns true if room is encrypted, false if not
     */
    static isRoomEncrypted(roomId: string): boolean {
        return MatrixClientPeg.get().isRoomEncrypted(roomId);
    }
}
