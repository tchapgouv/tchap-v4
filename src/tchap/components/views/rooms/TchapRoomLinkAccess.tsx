import { EventTimeline, EventType, GuestAccess, JoinRule, Room } from "matrix-js-sdk/src/matrix";

import React, { useEffect, useState } from 'react';
import LabelledToggleSwitch from '~tchap-web/src/components/views/elements/LabelledToggleSwitch';
import { _t } from '~tchap-web/src/languageHandler';
import { randomString } from 'matrix-js-sdk/src/randomstring';
import TchapRoomUtils from '../../../util/TchapRoomUtils';
import { makeRoomPermalink, RoomPermalinkCreator } from "~tchap-web/src/utils/permalinks/Permalinks";
import { TchapRoomType } from "../../../@types/tchap";
import { RoomJoinRulesEventContent } from "matrix-js-sdk/lib/types";
import CopyableText from "~tchap-web/src/components/views/elements/CopyableText";
import Modal from "~tchap-web/src/Modal";
import QuestionDialog from "~tchap-web/src/components/views/dialogs/QuestionDialog";
import ErrorDialog from "~tchap-web/src/components/views/dialogs/ErrorDialog";
import MatrixToPermalinkConstructor from "~tchap-web/src/utils/permalinks/MatrixToPermalinkConstructor";
import ElementPermalinkConstructor from "~tchap-web/src/utils/permalinks/ElementPermalinkConstructor";
import SdkConfig from "~tchap-web/src/SdkConfig";

interface ITchapRoomLinkAccessProps {
    room: Room,
    onUpdateParentView: Function
}

export default function TchapRoomLinkAccess({room, onUpdateParentView}: ITchapRoomLinkAccessProps) : JSX.Element {

    const [isLinkSharingActivated, setIsLinkSharingActivated] = useState(false);
    const [linkSharingUrl, setLinkSharingUrl] = useState("");
    const [disableLinkSharing, setDisableLinkSharing] = useState(false);

    // Getting the initial value of the link. We need to check if it was previsouly activated or not
    const initialLinkSharingValue = () => {

        // We disable link sharing if its a forum or user not admin
        if (!TchapRoomUtils.isUserAdmin(room) || TchapRoomUtils.getTchapRoomType(room) === TchapRoomType.Forum) {
            setDisableLinkSharing(true);
        }

        const isActivated = isJoinRulePublic();

        if (isActivated) {
            const link = makeRoomPermalink(room.client, room.roomId);
            setLinkSharingUrl(link);
        }
        setIsLinkSharingActivated(isActivated)
        // updating the parent join rule options 
        onUpdateParentView(isActivated, true);
    }

    useEffect(() => {
        initialLinkSharingValue();
    }, []);

    // Create the permalink to share
    const _setUpRoomByLink = async () => {
        try {
            let link = "";
            // create an alias if not existing
            if (!room.getCanonicalAlias()) {
                const aliasName = (room.name?.replace(/[^a-z0-9]/gi, "") ?? "") + randomString(11);
                const fullAlias = `#${aliasName}:${room.client.getDomain()}`;
                
                await room.client.createAlias(fullAlias, room.roomId)
                await room.client.sendStateEvent(room.roomId, EventType.RoomCanonicalAlias, { alias: fullAlias }, "")
                
                // we don't know why he cant detect directly that the canonical alias was created previously
                // so here we force the use of the newly created alias in the link
                const permalinkCreator = new RoomPermalinkCreator(room);
                permalinkCreator.load();
                const tchapPrefix: string = SdkConfig.get("permalink_prefix")!
                link = new ElementPermalinkConstructor(tchapPrefix).forRoom(fullAlias, permalinkCreator.serverCandidates || [])
            } else {
                // it will take the new alias created previously or the existing one to make a link
                link = makeRoomPermalink(room.client, room.roomId);
            }

            setLinkSharingUrl(link);
        } catch(err) {
            console.error(err);
            throw err;
        }
    };

    // Check if the current join rule is public or not
    const isJoinRulePublic = () => {
        const currentJoinRule: JoinRule = TchapRoomUtils.getRoomJoinRule(room) ?? JoinRule.Invite // if we dont receive the value we default to invite

        return currentJoinRule === JoinRule.Public
    }

    // Set the new join rule (public or invite)
    const _setJoinRules = (joinRule: JoinRule) => {
        return room.client.sendStateEvent(room.roomId, EventType.RoomJoinRules, { join_rule: joinRule } as RoomJoinRulesEventContent, "");
    };

    const _setGuestAccessRules = (guestAccess: GuestAccess) => {
        return room.client.sendStateEvent(room.roomId, EventType.RoomGuestAccess, {guest_access: guestAccess}, "")
    };
    
    // Handler to listen on the switch change 
    const _onLinkSharingSwitchChange = async (checked: boolean) => {
        try {
            let newJoinRule :JoinRule = checked ? JoinRule.Public : JoinRule.Invite;

            // if the link sharing is deactivated we also need to update the joinrule parent view to show the other options
            if (!checked) {
                await _setJoinRules(newJoinRule);
                onUpdateParentView(checked)
                setIsLinkSharingActivated(checked);
                return;
            }
    
            // Show modal for confirmation
            const activationIsConfirmed = await activateLinksharingModal();
    
            if (activationIsConfirmed) {
                // create link if we activate the sharing, otherwise change nothing
                if (TchapRoomUtils.getRoomGuessAccessRule(room) === GuestAccess.CanJoin) {
                    await _setGuestAccessRules(GuestAccess.Forbidden)
                }
                await Promise.all([_setUpRoomByLink(), _setJoinRules(JoinRule.Public)]);
                setIsLinkSharingActivated(checked);
                onUpdateParentView(checked, false);
            }
        } catch(err) {
            // we do nothing if there was an error
            console.error(err);
            Modal.createDialog(ErrorDialog, {
                title: _t("error_dialog|activate_room_share_link|title"),
                description: err && err.message ? err.message : _t("error_dialog|activate_room_share_link|description"),
            });
        }
    };


    const activateLinksharingModal = async (): Promise<boolean> => {
        const dialog = Modal.createDialog(QuestionDialog, {
            title: _t("room_settings|security|link_sharing_title"),
            description: (
                <div>
                    <p>
                        {_t("room_settings|security|link_sharing_modal_confirmation", null,   {
                        p: (sub) => <p>{sub}</p>,
                    },)}
                    </p>
                </div>
            ),
        });
        const { finished } = dialog;
        const [confirm] = await finished;
        return !!confirm
    }

    return (
        <div>
            <LabelledToggleSwitch value={isLinkSharingActivated}
                onChange={ _onLinkSharingSwitchChange }
                label={_t("room_settings|security|link_sharing_title")}
                caption={_t("room_settings|security|link_sharing_caption")}
                disabled={disableLinkSharing}
                data-testid="share_link_switch"
                />
            {
                isLinkSharingActivated ? 
                    <CopyableText getTextToCopy={() => linkSharingUrl} aria-labelledby="shared_room_link">
                        { linkSharingUrl }
                    </CopyableText>
                    : null
            }
        </div>
    )
}