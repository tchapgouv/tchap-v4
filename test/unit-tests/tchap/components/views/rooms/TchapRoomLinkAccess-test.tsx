import React from "react";
import { logRoles, render, screen, waitFor, waitForElementToBeRemoved } from "jest-matrix-react";
import { mocked } from "jest-mock";
import { EventType, GuestAccess, JoinRule, MatrixClient, Room } from "matrix-js-sdk/src/matrix";
import userEvent from "@testing-library/user-event";

import { flushPromises, waitEnoughCyclesForModal } from "../../../../../test-utils";

import { mkStubRoom, stubClient } from "~tchap-web//test/test-utils/test-utils";
import { makeRoomPermalink } from "~tchap-web//src/utils/permalinks/Permalinks";
import TchapRoomLinkAccess from "~tchap-web//src/tchap/components/views/rooms/TchapRoomLinkAccess";
import { TchapRoomType } from "~tchap-web//src/tchap/@types/tchap";
import TchapRoomUtils from "~tchap-web/src/tchap/util/TchapRoomUtils";
import SdkConfig, { ConfigOptions } from "~tchap-web/src/SdkConfig";

jest.mock("~tchap-web/src/tchap/util/TchapRoomUtils");
jest.mock("~tchap-web//src/utils/permalinks/Permalinks");
jest.mock("~tchap-web/src/utils/permalinks/ElementPermalinkConstructor");

describe("TchapRoomLinkAccess", () => {
    const client: MatrixClient = stubClient();
    const room: Room = mkStubRoom("roomId", "test", client);

    const mockedTchapRoomUtils = mocked(TchapRoomUtils);
    const mockedMakeRoomPermalink = mocked(makeRoomPermalink);
    const onUpdateParentView = jest.fn().mockImplementation(() => {});

    const mockedLinked = "https://testmocked.matrix.org";

    const getComponent = () => render(<TchapRoomLinkAccess room={room} onUpdateParentView={onUpdateParentView} />);

    const config: ConfigOptions = { permalink_prefix: "localhost" };
    SdkConfig.put(config);

    beforeEach(() => {
        mockedTchapRoomUtils.getTchapRoomType.mockImplementation(() => TchapRoomType.Private);
        mockedTchapRoomUtils.getRoomJoinRule.mockImplementation(() => JoinRule.Invite);
        mockedTchapRoomUtils.isUserAdmin.mockImplementation(() => true);
        mockedTchapRoomUtils.getRoomGuessAccessRule.mockImplementation(() => GuestAccess.CanJoin);
        mockedMakeRoomPermalink.mockImplementation(() => mockedLinked);

        client.createAlias = jest.fn().mockResolvedValue("alias");

        jest.spyOn(client, "sendStateEvent").mockResolvedValue(Promise.resolve({ event_id: "" }));
        jest.spyOn(client, "sendStateEvent").mockResolvedValue(Promise.resolve({ event_id: "" }));
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("should render correct initial value when joinrule is public", async () => {
        mockedTchapRoomUtils.getRoomJoinRule.mockImplementation(() => JoinRule.Public);
        mockedTchapRoomUtils.getTchapRoomType.mockImplementation(() => TchapRoomType.Forum);

        getComponent();

        await flushPromises();

        const switchLink = screen.queryByRole("switch");

        const linkDisplay = screen.queryByText(mockedLinked);

        expect(linkDisplay).toBeDefined();

        // should be disable because we dont change this settings for forum room
        expect(switchLink).toHaveAttribute("aria-disabled", "true");
    });

    it("should render correct initial value when joinrule is invite", async () => {
        mockedTchapRoomUtils.getRoomJoinRule.mockImplementation(() => JoinRule.Invite);

        getComponent();

        await flushPromises();

        const switchLink = screen.queryByRole("switch");

        const linkDisplay = screen.queryByText(mockedLinked);

        // linked should not appear because the share link is deactivated
        expect(linkDisplay).toBe(null);

        // we should be able to click on the link to activate it
        expect(switchLink).toHaveAttribute("aria-disabled", "false");
    });

    it("should disable link if user is not admin", async () => {
        mockedTchapRoomUtils.getRoomJoinRule.mockImplementation(() => JoinRule.Invite);
        mockedTchapRoomUtils.isUserAdmin.mockImplementation(() => false);

        getComponent();

        await flushPromises();

        const switchLink = screen.queryByRole("switch");

        const linkDisplay = screen.queryByText(mockedLinked);

        // linked should not appear because the share link is deactivated
        expect(linkDisplay).toBe(null);

        // the user should not be able to click on the button
        expect(switchLink).toHaveAttribute("aria-disabled", "true");
    });

    it("should activate link when clicking on the switch", async () => {
        mockedTchapRoomUtils.getRoomJoinRule.mockImplementation(() => JoinRule.Invite);

        getComponent();

        await flushPromises();

        const switchLink = screen.getByRole("switch");

        // should open dialog
        userEvent.click(switchLink);

        await waitEnoughCyclesForModal({
            useFakeTimers: true,
        });

        const dialog = await screen.findByRole("dialog");

        expect(dialog).toMatchSnapshot();

        const confirmButton = screen.getByTestId("dialog-primary-button");

        userEvent.click(confirmButton);

        await waitForElementToBeRemoved(dialog);
        // should activate the switch with public join rule value
        expect(room.client.createAlias).toHaveBeenCalledTimes(1);

        // joinrule to public, guest access to forbiden and canonical alias
        expect(room.client.sendStateEvent).toHaveBeenCalledTimes(3);

        await waitFor(() => {
            expect(switchLink).toHaveAttribute("aria-checked", "true");
        });
    });

    it("should use existing alias", async () => {
        jest.spyOn(room, "getCanonicalAlias").mockImplementation(() => "alias:domaing@test.fr");
        mockedTchapRoomUtils.getRoomJoinRule.mockImplementation(() => JoinRule.Invite);
        const { container } = getComponent();

        await flushPromises();

        const switchLink = screen.getByRole("switch");

        // should open dialog
        userEvent.click(switchLink);

        await waitEnoughCyclesForModal({
            useFakeTimers: true,
        });

        const dialog = await screen.findByRole("dialog");

        expect(dialog).toMatchSnapshot();

        const confirmButton = screen.getByTestId("dialog-primary-button");

        userEvent.click(confirmButton);

        await waitForElementToBeRemoved(dialog);
        // should activate the switch with public join rule value

        expect(mockedMakeRoomPermalink).toHaveBeenCalledTimes(1);

        logRoles(container);

        await waitFor(() => {
            expect(switchLink).toHaveAttribute("aria-checked", "true");
        });
    });

    it("should deactivate link when clicking on the switch", async () => {
        mockedTchapRoomUtils.getRoomJoinRule.mockImplementation(() => JoinRule.Public);

        getComponent();

        await flushPromises();

        const switchLink = screen.getByRole("switch");

        userEvent.click(switchLink);

        await waitFor(() => {
            const linkDisplay = screen.queryByText(mockedLinked);

            // no link since we deactivated the switch
            expect(linkDisplay).toBe(null);
            // should deactivate the switch
            expect(room.client.sendStateEvent).toHaveBeenCalledTimes(1);
            expect(room.client.sendStateEvent).toHaveBeenCalledWith(
                room.roomId,
                EventType.RoomJoinRules,
                { join_rule: JoinRule.Invite },
                "",
            );
            expect(switchLink).toHaveAttribute("aria-checked", "false");
        });
    });
});
