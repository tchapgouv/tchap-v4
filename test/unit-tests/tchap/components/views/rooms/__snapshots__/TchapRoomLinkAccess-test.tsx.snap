// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TchapRoomLinkAccess should activate link when clicking on the switch 1`] = `
<div
  aria-describedby="mx_Dialog_content"
  aria-labelledby="mx_BaseDialog_title"
  class="mx_QuestionDialog mx_Dialog_fixedWidth"
  data-focus-lock-disabled="false"
  role="dialog"
>
  <div
    class="mx_Dialog_header"
  >
    <h1
      class="mx_Heading_h3 mx_Dialog_title"
      id="mx_BaseDialog_title"
    >
      room_settings
    </h1>
  </div>
  <div
    class="mx_Dialog_content"
    id="mx_Dialog_content"
  >
    <div>
      <p>
        room_settings
      </p>
    </div>
  </div>
  <div
    class="mx_Dialog_buttons"
  >
    <span
      class="mx_Dialog_buttons_row"
    >
      <button
        data-testid="dialog-cancel-button"
        type="button"
      >
        Cancel
      </button>
      <button
        class="mx_Dialog_primary"
        data-testid="dialog-primary-button"
        type="button"
      >
        OK
      </button>
    </span>
  </div>
  <div
    aria-label="Close dialog"
    class="mx_AccessibleButton mx_Dialog_cancelButton"
    role="button"
    tabindex="0"
  />
</div>
`;
